teps/0033-tekton-feature-gates.md

```diff
@@ -145,6 +160,393 @@ alpha features. This is not a requirement for Tekton, because even in case of
+      implementation and use) than one flag per field
+    * Without adding feature gates, we risk either having to consider all new features immediately
+      beta/v1, or violating our own compatibility policies if changes are necessary
+* Flexibility:
+    * This proposal makes Tekton more flexible, at least for maintainers, because it makes it possible
+      to add new features and still have the ability to make updates to them in the future (without
+      waiting 6 or 9 months)
+* Conformance
+    * Fields should not be added to the Tekton conformance surface until they are promoted to the
+      same stability level as the CRD impacted
+
+## Drawbacks
+
+* Gating access to new features may decrease the number of adopters and make
+  it harder to get feedback
+
+## Alternatives
+
+### One feature flag per new Alpha field
+
+This alternative takes the same approach we propose for [changes in behavior](#changes-in-behavior)
+for [new API fields](#new-api-fields) as well: for each new field we add, we add a feature flag
+that gates use of that field.
+
+For example if we added a new feature to allow Tasks to fail in a Pipeline, and a feature to
+allow steps in a Task to fail, in order to enable both, the controller would need to be configured
+with:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: feature-flags
+data:
+  enable-allow-task-failure: "true"
+  enable-allow-step-failure: "true"
+```
+
+Pros:
+* Operators will have the ultimate control over what alpha features can be used
+
+Cons:
+* Users may have a harder time realizing what features they have available and we may
+  get less feedback (because they need to notice and enable each one)
+    * This could be mitigated by tooling
+* It will become harder to feel confident in the coverage of our tests. Features can
+  have unexpected side effects when combined (for example
+  [pipelines#3188](https://github.com/tektoncd/pipeline/issues/3188)
+  was a combination of when expressions and variable replacmement), but with just 3
+  different feature flags, we have 8 combinations of flags to potentially test
+  (no flags, just flag 1, flags 1 & 2, flags 1 & 3, just flag 2, flags 3 & 3,
+  just flag 3, all flags)
+
+### Maintain multiple verisons of impacted CRDs
+
+In this version, if we want to add an alpha field to a beta type, we would create
+a new alpha type and we would serve both.
+
+For example, right now Task has `v1alpha1` (deprecated) and `v1beta1`. We could also
+create a version such as `v1alpha2` and add new fields to that version before promoting them
+to `v1beta1`.
+
+Pros:
+* From the user's point of view, this might be the most intuitive option:
+  if they want to opt into alpha features for a beta CRD (e.g. Task is currently v1beta1),
+  they use the alpha version (v1alpha1)
+* Operators can control access by not installing CRD versions they don't want being used
+
+Cons:
```

Comment:
There is one additionnal `con` here: in Kubernetes, you can have only 1 version of the CRD stored, and all the rest are converted to that version. So you need to decide which one it will be, and how to re-hydrate different field from a given version to another one. This is *mainly* a con because of our crd+webhook+controller approach. Having a full-fledge apiserver (using api aggregation, …) would allow us to maintain a stored version completely out of the one we publish.